### PR TITLE
Add guardrails to prevent and correct wrong output from JavaTypeCorrect

### PR DESCRIPTION
**Pull request by kelloggm**
~~Incorporates #338, so don't review until that is merged.~~

~~I'm not sure if we want to merge this. It avoids some bad behavior on the NA-176 test, and we could probably expand it in the future for other, similar kinds of problems. However, it definitely does not address the root cause: why is Specimin putting itself in a situation where javac insists that a synthetic class must be compatible with String?~~

~~I'm going to continue my investigation, but I'm leaving this PR here because I may want to merge it later if the fix for the root cause is too complex and/or impracticle.~~

This PR makes a series of changes to how our javac type correction algorithm (in the `JavaTypeCorrect` class) handles complex cases. That algorithm is fundamentally imprecise, so the changes in this PR should be thought of as incremental improvements to the existing design that avoid some of its worst pathologies. In particular, this PR:
* prevents JavaTypeCorrect from ever attempting to make a class extend `java.lang.String`, which is final;
* makes it generally more difficult for the `SyntheticUnconstrainedType` type variable to propagate beyond the return type to which it is applied; and,
* replaces parameters whose types are synthetic return types that get replaced by `SyntheticUnconstrainedType` with `java.lang.Object`. Previously, these types would hang around and make the output non-compilable, since they'd been marked as unused.

Together, these changes make NullAway-176 reproducible, based on my local testing.

### CODE REVIEW

**Review by theron-wang — APPROVED** [review]
Looks good to me, aside from one minor comment!

File: src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java [hunk 1]
```diff
@@ -104,6 +107,24 @@ public Map<String, String> getClassAndUnresolvedInterface() {
    * @return the map described above.
    */
   public Map<String, String> getExtendedTypes() {
+    // Before returning, purge any entries that are obviously bad according to
+    // the following simple heuristic(s):
+    // * don't extend known-final classes from the JDK, like java.lang.String.
```
[the remark this answers is not in the document]
Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java:112]: I thought we might want to add more in the future, so I future-proofed the comment. Your point is a good one, though - what we should do is look up a list, and then add that to `JavaLangUtils`. I'll do that now.

File: src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java [hunk 2]
```diff
@@ -114,7 +114,7 @@ public Map<String, String> getExtendedTypes() {
     Set<String> toRemove = new HashSet<>(0);
     for (Map.Entry<String, String> entry : extendedTypes.entrySet()) {
       // String is final
```
Comment (theron-wang) [inline src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java:116]: This comment can be removed/adjusted

**Review by theron-wang — APPROVED** [review]
LGTM, one small comment change.

**Pull request merged**